File: leetcode/34/SearchRange.java
import java.util.Arrays;

class SearchRange {
    public static void main(String[] args) {
        Solution sl = new Solution();
        System.out.println(Arrays.toString(sl.searchRange(new int[] { 1, 2, 3, 4, 5, 7, 7, 7, 7, 8 }, 7)));
    }
}
class Solution {
    public int[] searchRange(int[] nums, int target) {
        return new int[] { findFirst(nums, target), findLast(nums, target) };
    }

    public int findFirst(int[] nums, int target) {
        int index = -1;

        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] >= target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
            if (nums[mid] == target)
                index = mid;
        }

        return index;
    }

    public int findLast(int[] nums, int target) {
        int index = -1;

        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] <= target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
            if (nums[mid] == target)
                index = mid;
        }

        return index;
    }
}

File: leetcode/34/SearchRange.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    // let result = [];
    // result[0] = findFirst(nums,target);
    // result[1] = findLast(nums,target);
    // return result;
    return [findFirst(nums,target),findLast(nums,target)];
    
};

var findFirst = function(nums, target) {
    let index = -1;

    let start = 0;
    let end = nums.length - 1;
    
    while (start <= end) {
        let mid = start + Math.floor((end - start) / 2);
        if (nums[mid] >= target) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
        if (nums[mid] == target)
            index = mid;
    }
    
    return index;
};

var findLast = function(nums, target) {
    let index = -1;

    let start = 0;
    let end = nums.length - 1;

    while (start <= end) {
        let mid = start + Math.floor((end - start) / 2);
        if (nums[mid] <= target) {
            start = mid + 1;
        } else {
            end = mid - 1;
        }
        if (nums[mid] == target)
            index = mid;
    }

    return index;
};